### ISSUE TITLE
Consider dropping Python dependency

### ISSUE THREAD

**Issue by alexeagle** [keith/rules_multirun#43]
rules_python isn't a great dependency. It's still pre-1.0 and not very stable.

If rules_multirun is used beneath some ruleset, the rules_python dependency that ruleset takes can easily replace the user's, leading to breaking changes. This is true under both WORKSPACE and bzlmod.

https://www.gnu.org/software/parallel/ is a good alternative that can be installed hermetically and doesn't require any runtime.

fyi @thesayyn who might be able to contribute it, if you like the idea :)

**Comment by keith** [comment]
I'm open to something, but it seems tough because it's all about the tradeoffs. The original version of this tool was using golang, but I didn't want to force that toolchain on folks either. And if we use more shell we might have that same kind-of issue with windows (I'm still not sure how big of an issue that is for windows support). Given how widely used rules_python is I'm not particularly phased by the version number, but I can't blame people if they don't otherwise use python and now depend on that through this

**Comment by alexeagle** [comment]
FWIW we use Golang for tools in our bazel repos, but include pre-built binaries on releases so users don't have any toolchain dependency. It's pretty conveniently tucked into our release automation so it doesn't cost us much. https://blog.aspect.dev/releasing-bazel-rulesets-rust

**Comment by keith** [comment]
yea i think doing something like that could be a great solution

**Comment by alexeagle** [comment]
Note we also have a nice pattern for distributing Rust tools, if the eventual contributor here prefers

**Comment by peakschris** [comment]
@alexeagle are you familiar with the go multirun that this repo was derived from? It's still under development although the owner doesn't make releases. I have been wondering if combining the best of both, ending up with a go solution that can be released as binary, would be a nice idea. It would need to have a formal release mechanism for us to depend on it.

https://github.com/ash2k/bazel-tools

**Comment by keith** [comment]
yea i think that, avoiding depending on the entire go world, could be ideal. alternatively we could just use go for building and vendor the binaries for releases as some repos do

**Comment by alexeagle** [comment]
I can contribute the bits to use rules_go here as a devDependency. I wonder if that Go implementation is still compatible? And if we switch to it, maybe we should just go back to that fork

**Comment by peakschris** [comment]
I've been doing some trial work on the go multirun. It's looking promising, most test cases are now passing on windows with runfiles on and the underlying functionality also works with runfiles off. I think the API is largely the same as here.

I like the idea of keeping this rules_multitool repo, it's lean. I could contribute a PR that added windows support to the go implementation and could help out in to the port effort. If desired, I could also do a draft PR that added the go code to this repo.

I can't be an ongoing maintainer, I'm in the middle of porting a huge monorepo to bazel and won't have much time once I've got rules_lint working for us (that's my main driver right now)

Which CI / binary build infra would be used?

**Comment by alexeagle** [comment]
https://blog.aspect.build/releasing-bazel-rulesets-rust describes the build infra that has the desired properties:
- release just by pushing a tag
- mark rules_go as a devDependency when users are on a release build

It's minimal effort for the rules maintainer and doesn't leak your tooling into users repos.

**Comment by peakschris** [comment]
That looks good, Alex.

I've spent some time working on the go version's code. I was trying to update the go repository test cases to support windows and norunfiles, and it's got too complex; the combination of a lot of different ways to pass arguments and quote them has made the complexity skyrocket to a point where I think it would be unmaintainable.

My conclusion is that it would be better to stick with Keith's implementations; the API is tidier, and smaller, and so the test cases and usage patterns out in the wild are likely to be simpler. 

I think it would be good to:
* Bring across and integrate the go runner and modify it to work with keith's existing rules
* Update the rules to support bat wrappers on windows
* Add go binary builds/distribution

**Comment by peakschris** [comment]
@keith @alexeagle Please see https://github.com/keith/rules_multirun/pull/57, I'm keen to have your feedback and ideas. It's a big change - needs careful discussion.

**Comment by alexeagle** [comment]
Sorry there's no time! Hair on fire! but I really appreciate your work on that and making Bazel better on Windows. Please keep pushing on us :)

**Comment by peakschris** [comment]
Alex, no worries - I know the feeling, and I'll keep reminding for when you have a chance!

Just to update, there is a chain of PRs that I need to get in to fix various windows issues we are facing, and to enable use of rules_lint format command:
- 1: https://github.com/aspect-build/bazel-lib/pull/872
- 2: https://github.com/keith/rules_multirun/pull/57
- 3: aspect_lint - update to rules_multirun to fix issues with format on windows

@keith could you let us know your thoughts?

**Comment by keith** [comment]
I'm generally supportive here, I just need some time to review that one since it's pretty big

**Comment by ash2k** [comment]
Hah, I have just discovered this issue :) I'm the original author of multirun. It was first written in shell/bash but then I couldn't make it handle some OS signals and rewrote in Go since that's what I used already. I've since switched to the implementation in this repo.

FWIW I'm also not a fan of the Python dependency, but it works for now.

**Comment by pnathan** [comment]
I am using this in a remote build execution system, and rules_multirun is crashing due to some python3 not found issue. I would vastly prefer a golang/vendored binary solution that can pack everything tidily up without relying on host rightness.

**Comment by keith** [comment]
Note that the expectation is that you're using a hermetic toolchain from rules_python, so it sounds like there's a different issue in your project

**Comment by albertocavalcante** [comment]
Is it safe to assume switching from `Python` to native binaries it the likely outcome of this issue and you're supportive @Keith?
Any thoughts to consider? Thank you!

**Comment by keith** [comment]
im not sure, here's an example of that https://github.com/keith/rules_multirun/pull/57 and there's some complexity to consider. are there specific issues with the python implementation and hermetic python toolchains you're seeing?

**Comment by ash2k** [comment]
Yeah, I'm now having issues using `rules_python` without system Python installed. See [this](https://github.com/bazel-contrib/rules_python/blob/main/BZLMOD_SUPPORT.md#default-toolchain-is-not-the-local-system-python):

> NOTE: Regardless of your toolchain, due to https://github.com/bazel-contrib/rules_python/issues/691, rules_python still relies on a local Python being available to bootstrap the program before handing over execution to the toolchain Python.

The issue it links to is such a rabbit hole. I'm too getting `/usr/bin/env: 'python3': No such file or directory` in my build 😕 Having to have Python to bootstrap Python is not great.

I saw that bazel 8.3.0 has wasm support now. This looks very exciting and promising for solving these bootstrap issues. See https://github.com/bazelbuild/bazel/pull/26280. I wonder if we could build a wasm binary that can be used instead of Python and drop `rules_python`? It should work on any OS without any dependencies, right?

**Comment by keith** [comment]
ah yea that's a bummer. i think they plan to flip the default of `--bootstrap_impl` to fix that soon, but I'm not sure what's blocking that or if it entirely solves it.

The wasm stuff is definitely interesting, although i think the solution for that vs distributing N native binaries would end up looking pretty similar wrt complexity of distributing things with the release

**Comment by peakschris** [comment]
Bazel's python support is so frustrating -- but also required so widely. Hermetic python doesn't work on large repos on windows (performance problems). After holding out for a long time we are having to bite the bullet and standardize on a specific system python version for every OS in order to provide it to bazel and various rulesets that require it.

**Comment by alexeagle** [comment]
@peakschris I'm still really keen on getting a Go binary into this repo. I'll help with the release automation if we want users to get a prebuilt binary rather than build it from source. Do you want to take another try at your PR to merge the python impl here with the original from @ash2k (https://github.com/ash2k/bazel-tools/tree/415483a9e13342a6603a710b0296f6d85b8d26bf/multirun)

**Comment by alexeagle** [comment]
Testing at HEAD

```
git_override(
    module_name = "rules_multirun",
    commit = "e2245c6a4e1fb26cea7db6fac16d856693cf1ec1",
    remote = "https://github.com/keith/rules_multirun.git",
)
```

I have `common --@rules_python//python/config_settings:bootstrap_impl=script` in .bazelrc so other console_script's are working on a machine with no python installed.

```
$ copier --help
rules_python_entry_point_copier.py 9.10.2

Create a new project from a template.
...
```

But running a multirun target (via rules_lint format_multirun)

```
$ aspect format
...
ERROR: Python interpreter not found: /workspaces/py/bazel-out/bazel_env-opt/bin/tools/bazel_env/bin/aspect.runfiles/rules_multirun+/internal/_multirun.venv/bin/python3
ls: cannot access '/workspaces/py/bazel-out/bazel_env-opt/bin/tools/bazel_env/bin/aspect.runfiles/rules_multirun+/internal/_multirun.venv/bin/python3': No such file or directory
```
